pyomo/contrib/mindtpy/util.py

```diff
@@ -180,3 +181,156 @@ def var_bound_add(solve_data, config):
+
+    Parameters
+    ----------
+    model: Pyomo model
+        the model that needs new objective function
+    setpoint_model: Pyomo model
+        the model that provides the base point for us to calculate the distance
+    discrete_only: Bool
+        only optimize on distance between the discrete variables
+    TODO: remove setpoint_model
+    """
+    var_filter = (lambda v: v[1].is_integer()) if discrete_only \
+        else (lambda v: True)
+
+    model_vars, setpoint_vars = zip(*filter(var_filter,
+                                            zip(model.component_data_objects(Var),
+                                                setpoint_model.component_data_objects(Var))))
+    assert len(model_vars) == len(
+        setpoint_vars), "Trying to generate Squared Norm2 objective function for models with different number of variables"
+
+    return Objective(expr=(
+        sum([(model_var - setpoint_var.value)**2
+             for (model_var, setpoint_var) in
+             zip(model_vars, setpoint_vars)])))
+
+
+def generate_norm1_objective_function(model, setpoint_model, discrete_only=False):
+    """
+    This function generates objective (PF-OA master problem) for minimum Norm1 distance to setpoint_model
+    Norm1 distance of (x,y) = \sum_i |x_i - y_i|
+
+    Parameters
+    ----------
+    model: Pyomo model
+        the model that needs new objective function
+    setpoint_model: Pyomo model
+        the model that provides the base point for us to calculate the distance
+    discrete_only: Bool
+        only optimize on distance between the discrete variables
+    TODO: remove setpoint_model
+    """
+
+    var_filter = (lambda v: v.is_integer()) if discrete_only \
+        else (lambda v: True)
+    model_vars = list(filter(var_filter, model.component_data_objects(Var)))
+    setpoint_vars = list(
+        filter(var_filter, setpoint_model.component_data_objects(Var)))
+    assert len(model_vars) == len(
+        setpoint_vars), "Trying to generate Norm1 objective function for models with different number of variables"
+    if model.MindtPy_utils.find_component('L1_objective_function') is not None:
+        model.MindtPy_utils.del_component('L1_objective_function')
+    obj_blk = model.MindtPy_utils.L1_objective_function = Block()
+    obj_blk.L1_obj_idx = RangeSet(len(model_vars))
+    obj_blk.L1_obj_var = Var(
+        obj_blk.L1_obj_idx, domain=Reals, bounds=(0, None))
+    obj_blk.abs_reformulation = ConstraintList()
+    for idx, v_model, v_setpoint in zip(obj_blk.L1_obj_idx, model_vars,
+                                        setpoint_vars):
+        obj_blk.abs_reformulation.add(
+            expr=v_model - v_setpoint.value >= -obj_blk.L1_obj_var[idx])
+        obj_blk.abs_reformulation.add(
+            expr=v_model - v_setpoint.value <= obj_blk.L1_obj_var[idx])
+
+    return Objective(expr=sum(obj_blk.L1_obj_var[idx] for idx in obj_blk.L1_obj_idx))
+
+# TODO: this function is not called
```

Comment:
Which function?

Reply:
`generate_norm_inf_objective_function()`

Reply:
We should add a test for this function similar to the L2 norm test